zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/util/ZkPathRecursiveWatcherTrie.java

```diff
@@ -139,17 +139,88 @@ public void addRecursiveListener(final String path, RecursivePersistListener lis
     if (path.isEmpty()) {
       throw new IllegalArgumentException("Empty path: " + path);
     }
-    final List<String>  pathComponents = split(path);
+    final List<String> pathComponents = split(path);
 
-    synchronized(this) {
+    synchronized (this) {
       TrieNode parent = _rootNode;
       for (final String part : pathComponents) {
-        parent = parent.getChildren().computeIfAbsent(part, (p)-> new TrieNode(part) );
+        parent = parent.getChildren().computeIfAbsent(part, (p) -> new TrieNode(part));
       }
       parent._recursiveListeners.add(listener);
     }
   }
 
+  public Set<RecursivePersistListener> getAllRecursiveListeners(String path) {
+    Objects.requireNonNull(path, "Path cannot be null");
+
+    final List<String> pathComponents = split(path);
+    Set<RecursivePersistListener> result = new HashSet<>();
+    synchronized (this) {
+      TrieNode cur = _rootNode;
+      for (final String element : pathComponents) {
+        cur = cur.getChild(element);
+        if (cur == null) {
+          break;
+        }
+        result.addAll(cur.getRecursiveListeners());
+      }
```

Comment:
This piece of code can be generalized:

synchronized (this) {
      TrieNode cur = _rootNode;
      for (final String element : pathComponents) {
        cur = cur.getChild(element);
        if (cur == null) {
          break;
        }
        callable.call(cur);
      }


For this call() {
  result.addAll(cur.getRecursiveListeners());
}

Reply:
Had an offline discussion. It is hard to generalize this part of logic with the following func.